File: src/parser_utils/output/printers/compact_printer.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Layout

let print node =
  let rec print_node src = function
    | SourceLocation (_loc, node) -> print_node src node
    | Concat nodes
    | Group nodes
    | Sequence (_, nodes) ->
      List.fold_left print_node src nodes
    | Indent node -> print_node src node
    | Newline -> Source.add_newline src
    | Atom s -> Source.add_string s src
    | Identifier (_loc, s) -> Source.add_identifier s src
    | IfPretty (_, node) -> print_node src node
    | IfBreak (_, otherwise) -> print_node src otherwise
    | Empty -> src
  in
  let src = print_node (Source.create ()) node in
  Source.add_newline src

File: rust_port/crates/flow_parser_utils_output/src/compact_printer.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use crate::layout::LayoutNode;
use crate::layout::LayoutNodeInner;
use crate::source::Source;

pub fn print(node: &LayoutNode) -> Source {
    fn print_node(src: &mut Source, node: &LayoutNode) {
        match node.inner() {
            LayoutNodeInner::SourceLocation(_loc, node) => print_node(src, node),
            LayoutNodeInner::Concat(nodes)
            | LayoutNodeInner::Group(nodes)
            | LayoutNodeInner::Sequence(_, nodes) => {
                for node in nodes.iter() {
                    print_node(src, node);
                }
            }
            LayoutNodeInner::Indent(node) => print_node(src, node),
            LayoutNodeInner::Newline => {
                src.add_newline();
            }
            LayoutNodeInner::Atom(s) => {
                src.add_string(s);
            }
            LayoutNodeInner::Identifier(_loc, s) => {
                src.add_identifier(s);
            }
            LayoutNodeInner::IfPretty(_, node) => print_node(src, node),
            LayoutNodeInner::IfBreak(_, otherwise) => print_node(src, otherwise),
            LayoutNodeInner::Empty => {}
        }
    }
    let mut src = Source::default();
    print_node(&mut src, node);
    src.add_newline();
    src
}
